### PR TITLE
Migrate gallery ios tests to build+test

### PR DESCRIPTION
**Pull request by keyonghan**
Part of https://github.com/flutter/flutter/issues/103542

### CODE REVIEW

File: dev/devicelab/lib/tasks/gallery.dart [hunk 1]
```diff
@@ -359,7 +378,12 @@ class GalleryTransitionBuildTest extends BuildTestTask {
   @override
   String getApplicationBinaryPath() {
     if (applicationBinaryPath != null) {
-      return '${applicationBinaryPath!}/app-profile.apk';
+      if(deviceOperatingSystem == DeviceOperatingSystem.android) {
+        return '$applicationBinaryPath/app-profile.apk';
+      }
+      else if (deviceOperatingSystem == DeviceOperatingSystem.ios) {
+        return '$applicationBinaryPath';
```
Comment (keyonghan) [inline dev/devicelab/lib/tasks/gallery.dart:385]: @jmagman is this the correct path forwarded to test step as `'--use-application-binary=${getApplicationBinaryPath()}',`?
I somehow hit the error complaining [`it's not an app bundle`](https://logs.chromium.org/logs/flutter/led/keyonghan_google.com/00f9030f3e207c0f45e97866bfdb68dd32647102e16923278869df62b7d4d4f9/+/u/run_flutter_gallery__transition_perf_e2e_ios/test_stdout)

What I am doing is:
* build the artifacts (upload `dev/integration_tests/flutter_gallery/build/ios/iphoneos/Flutter Gallery.app/*` to [gcs](https://pantheon.corp.google.com/storage/browser/flutter_devicelab/flutter/staging/refs/pull/111164/head/gallery__transition_perf_e2e_ios?project=flutter-infra-staging))
* download those artifacts to `<temp dir>/gallery__transition_perf_e2e_ios` and inject that path to `--use-application-binary`.

Comment (jmagman) [reply dev/devicelab/lib/tasks/gallery.dart:385]: I think it wants the specific `.app` in the directory `--use-application-binary=/opt/s/w/ir/x/w/recipe_cleanup/tmpvx23us9j/gallery__transition_perf_e2e_ios`

File: dev/devicelab/lib/tasks/gallery.dart [hunk 2]
```diff
@@ -251,25 +251,46 @@ class GalleryTransitionBuildTest extends BuildTestTask {
 
   @override
   void copyArtifacts() {
-    if(applicationBinaryPath != null) {
+    if(applicationBinaryPath == null) {
```
Comment (jmagman) [inline dev/devicelab/lib/tasks/gallery.dart:254]:
```suggestion
    if (applicationBinaryPath == null) {
```

File: dev/devicelab/lib/tasks/gallery.dart [hunk 3]
```diff
@@ -251,25 +251,46 @@ class GalleryTransitionBuildTest extends BuildTestTask {
...
+      return;
+    }
+    if(deviceOperatingSystem == DeviceOperatingSystem.android) {
```
Comment (jmagman) [inline dev/devicelab/lib/tasks/gallery.dart:257]:
```suggestion
    if (deviceOperatingSystem == DeviceOperatingSystem.android) {
```

File: dev/devicelab/lib/tasks/gallery.dart [hunk 4]
```diff
@@ -251,25 +251,46 @@ class GalleryTransitionBuildTest extends BuildTestTask {
...
       copy(
         file('${galleryDirectory.path}/build/app/outputs/flutter-apk/app-profile.apk'),
         Directory(applicationBinaryPath!),
       );
+    } else if(deviceOperatingSystem == DeviceOperatingSystem.ios) {
```
Comment (jmagman) [inline dev/devicelab/lib/tasks/gallery.dart:262]:
```suggestion
    } else if (deviceOperatingSystem == DeviceOperatingSystem.ios) {
```

File: dev/devicelab/lib/tasks/gallery.dart [hunk 5]
```diff
@@ -359,7 +380,12 @@ class GalleryTransitionBuildTest extends BuildTestTask {
   @override
   String getApplicationBinaryPath() {
     if (applicationBinaryPath != null) {
-      return '${applicationBinaryPath!}/app-profile.apk';
+      if(deviceOperatingSystem == DeviceOperatingSystem.android) {
```
Comment (jmagman) [inline dev/devicelab/lib/tasks/gallery.dart:383]:
```suggestion
      if (deviceOperatingSystem == DeviceOperatingSystem.android) {
```

File: .ci.yaml [hunk 6]
```diff
@@ -3706,6 +3706,7 @@ targets:
       tags: >
         ["devicelab", "ios", "mac"]
       task_name: flutter_gallery__transition_perf_e2e_ios
+      artifact: gallery__transition_perf_e2e_ios
```
Comment (jmagman) [inline .ci.yaml:3709]: These artifacts can't be shared, `flutter_gallery__transition_perf_e2e_impeller_ios` it built with impeller and `flutter_gallery__transition_perf_e2e_ios` is not.

It seems like we should have some more generic way to do this, like hashing the example/integration_test app directory and the build flags?  Also, some tests build the app multiple times with different flags (debug, profile, release).

File: dev/devicelab/lib/tasks/gallery.dart (hunk 1 above)
Comment (keyonghan) [reply dev/devicelab/lib/tasks/gallery.dart]: This is it!

File: .ci.yaml (hunk 6 above)
Comment (keyonghan) [reply .ci.yaml:3709]: Right, different tasks with different building args are not sharing these artifact.
I am planning to drop this `artifact` property for non-sharing ones. If two tasks do share the same artifact, we will specify the artifact explicitly.

Comment (keyonghan) [reply .ci.yaml]: Latest change: removed the artifact property here. In general, if `artifact` is not specified, `task name` will be used as `artifact` for the task. On the other side, if specified, the task will use it, and other task will share it if same `artifact` is defined.

This way, we have an assumption that each task builds a single artifact. This way, tasks that follow build+test+build+test etc. will not be supported for the later artifact sharing. As talked in go/devicelab-build-test, the later build should be covered in the `test` step.

Please let me know if this makes sense to you.

File: .ci.yaml [hunk 7]
```diff
@@ -156,6 +156,17 @@ platform_properties:
...
+    properties:
+      dependencies: >-
+        [
+          {"dependency": "xcode", "version": "14c18"},
+          {"dependency": "gems", "version": "v3.3.14"},
+          {"dependency": "apple_signing", "version": "version:2022_to_2023"}
+        ]
+      os: Mac-12
+      device_type: none
+      xcode: 14c18 # Xcode 14.2 to support iOS 16.2 in devicelab.
```
Comment (jmagman) [inline .ci.yaml:169]: How are the recipes chosen?  If it's devicelab this is fine since @godofredoc has said it has hermetic Xcode caches, and it seems like the `Mac_build_test` tests are going to need tethered devices.  However if it's not the devicelab recipe then it should be

```suggestion
      xcode: 14a5294e # xcode 14.0 beta 5
```

because of the Xcode cache issues: https://github.com/flutter/flutter/issues/117541#issuecomment-1363232756

@godofredoc can you clarify if this is right?

File: .ci.yaml [hunk 8]
```diff
@@ -4033,6 +4044,18 @@ targets:
...
+    recipe: devicelab/devicelab_drone_build_test
+    bringup: true # New target: https://github.com/flutter/flutter/pull/111164
+    presubmit: false
+    timeout: 60
+    properties:
+      tags: >
+        ["devicelab", "ios", "mac"]
+      task_name: flutter_gallery__transition_perf_e2e_ios
+      drone_dimensions: >
+        ["device_os=iOS-16","os=Mac-12", "cpu=x86"]
```
Comment (jmagman) [inline .ci.yaml:4057]: Why is this necessary if it's `mac_build_test`?
If it is necessary, should they be properties?
```
      os: Mac-12
      cpu: x86
      device_os: iOS-16
```

File: dev/devicelab/lib/tasks/gallery.dart [hunk 9]
```diff
@@ -251,25 +251,44 @@ class GalleryTransitionBuildTest extends BuildTestTask {
...
+      return;
+    }
+    if (deviceOperatingSystem == DeviceOperatingSystem.android) {
       copy(
         file('${galleryDirectory.path}/build/app/outputs/flutter-apk/app-profile.apk'),
         Directory(applicationBinaryPath!),
       );
+    } else if (deviceOperatingSystem == DeviceOperatingSystem.ios) {
+      recursiveCopy(
+        Directory('${galleryDirectory.path}/build/ios/iphoneos'),
```
Comment (jmagman) [inline dev/devicelab/lib/tasks/gallery.dart:264]: And copying only this directory prevents a rebuild?

File: dev/devicelab/lib/tasks/gallery.dart [hunk 10]
```diff
@@ -359,7 +378,12 @@ class GalleryTransitionBuildTest extends BuildTestTask {
...
-      return '${applicationBinaryPath!}/app-profile.apk';
+      if (deviceOperatingSystem == DeviceOperatingSystem.android) {
+        return '$applicationBinaryPath/app-profile.apk';
+      }
+      else if (deviceOperatingSystem == DeviceOperatingSystem.ios) {
+        return '$applicationBinaryPath/Flutter Gallery.app';
+      }
     }
 
     return 'build/app/outputs/flutter-apk/app-profile.apk';
```
Comment (jmagman) [inline dev/devicelab/lib/tasks/gallery.dart:389]: Should this also be dependent on the `DeviceOperatingSystem`?

Comment (keyonghan) [reply dev/devicelab/lib/tasks/gallery.dart:389]: We currently support only android and ios. The return is the expected binary path, which already depends on the device os. Updated the logic a bit to make it clear.

File: dev/devicelab/lib/tasks/gallery.dart (hunk 9 above)
Comment (keyonghan) [reply dev/devicelab/lib/tasks/gallery.dart:256]: Yes. Validated using LED runs. Example: https://luci-milo.appspot.com/raw/build/logs.chromium.org/flutter/led/keyonghan_google.com/a43684afc067cba831ca3268c9621008cae1b916634886dada4c5c4a8c3d3587/+/build.proto, which skipped the `build` step, and ran test based on earlier built artifact.

File: .ci.yaml (hunk 7 above)
Comment (keyonghan) [reply .ci.yaml]: Good catch. The parent build is running on a host only bot, and should use same xcode as other host only targets. Updated to use 14a5294e and added comments to clarify

Comment (keyonghan) [reply .ci.yaml]: The `devicelab_drone_build_test` recipe currently injects the `devicelab_osx_sdk` version based on property `xcode` to run `test`. This will cause the `test` run on a `devicelab` bot but with an `xcode` version defined for the `build` running on the host only bot. This is clearly not expected.

Based on current logics (ci.yaml roller & cocoon scheduler) which auto generate targets' `xcode` (`devicelab_osx_sdk` or `osx_sdk`) based on the platform `xcode`, this PR as of now will not work.

**We expect:**

- `osx_sdk: 14a5294e` for build step on the host only bot
- `devicelab_osx_sdk: 14c18` for test step on the devicelab bot.

**We get:**

- same version for both.

To make this PR work, one way we can do is twist both the `ci.yaml roller` and `cocoon scheduler` to add both `osx_sdk` and `devicelab_osx_sdk` to the target. But this will add additional complexity to existing logics.

On the other hand, I do not see we will have a short-term fix on the xcode issue. To make things easier and unblock here. I am thinking to add `clean cache` flag to host only bots, allowing bots a couple of days to clean up old xcode's cache and prepare for the new version. This way, we will be unblocked here for all devicelab mitgration to build+test, and also make xcode version consistent across host only and devicelab targets. 

With that being said, and considering ongoing frequent xcode changes, we should this to Q2.

@godofredoc before the xcode issue resolved from root (hopefully in Q2), do you have any objection for the above proposal?

**Comment by jmagman** [comment]
Swapping out myself as a reviewer for @christopherfujino.

File: .ci.yaml (hunk 7 above)
Comment (keyonghan) [reply .ci.yaml]: This is unblocked, and have updated to use the latest xcode: 14e222b.

First run when no artifact exists: https://luci-milo.appspot.com/raw/build/logs.chromium.org/flutter/led/keyonghan_google.com/abc2f7c6ea7e36380cb59c73527839a5f35bd9ea5e51ce7c0d210fe704957246/+/build.proto
Second run when the artifact exists: https://luci-milo.appspot.com/raw/build/logs.chromium.org/flutter/led/keyonghan_google.com/5fe9cb5279aa53d509bb1cce94491cc5911bd9873851302b180e7205de244cb8/+/build.proto

@christopherfujino Please help take a look as @jmagman is ooo. Thanks!

**Comment by christopherfujino** [comment]
@keyonghan is this ready for review?

**Comment by keyonghan** [comment]
> @keyonghan is this ready for review?

Yes. Please help take a look. (sorry about the noise to restore the format)

File: dev/devicelab/bin/tasks/flutter_gallery__transition_perf_e2e_ios.dart [hunk 11]
```diff
@@ -6,7 +6,7 @@ import 'package:flutter_devicelab/framework/devices.dart';
 import 'package:flutter_devicelab/framework/framework.dart';
 import 'package:flutter_devicelab/tasks/gallery.dart';
 
-Future<void> main() async {
+Future<void> main(List<String> args) async {
   deviceOperatingSystem = DeviceOperatingSystem.ios;
-  await task(createGalleryTransitionE2ETest());
+  await task(createGalleryTransitionE2EBuildTest(args));
```
Comment (christopherfujino) [inline dev/devicelab/bin/tasks/flutter_gallery__transition_perf_e2e_ios.dart:11]: do we ever use these args?

Comment (keyonghan) [reply dev/devicelab/bin/tasks/flutter_gallery__transition_perf_e2e_ios.dart:11]: We do. See `GalleryTransitionBuildTest` extends `BuildTestTask`, which parses these args. For context, this is how these args look like: https://flutter.googlesource.com/recipes/+/refs/heads/main/recipes/devicelab/devicelab_drone_build_test.py#142

Comment (christopherfujino) [reply dev/devicelab/bin/tasks/flutter_gallery__transition_perf_e2e_ios.dart:11]: I see. so we were never parsing these args before?

Comment (keyonghan) [reply dev/devicelab/bin/tasks/flutter_gallery__transition_perf_e2e_ios.dart:11]: That's correct. These args are used for build+test tasks only.

Btw: `Linux gallery` tasks have been running in prod for months, example: `Linux_build_test flutter_gallery__transition_perf`

**Review by christopherfujino — APPROVED** [review]
LGTM
